In [172]:
import pandas as pd
from pathlib import Path
from datetime import datetime

In [173]:
csv_path = Path("./../../../data/clean/Yashaswi/Atlantic_Climate.csv")
df = pd.read_csv(csv_path, low_memory=False)

In [174]:
df.info()
df.head()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Longitude (x)              964749 non-null  float64
 1   Latitude (y)               964749 non-null  float64
 2   Station Name               964749 non-null  str    
 3   Climate ID                 964749 non-null  str    
 4   Date/Time                  964749 non-null  str    
 5   Year                       964749 non-null  int64  
 6   Max Temp (°C)              828481 non-null  float64
 7   Min Temp (°C)              831117 non-null  float64
 8   Mean Temp (°C)             826411 non-null  float64
 9   Heat Deg Days (°C)         826411 non-null  float64
 10  Cool Deg Days (°C)         826411 non-null  float64
 11  Total Rain (mm)            169711 non-null  float64
 12  Total Snow (cm)            165870 non-null  float64
 13  Total Precip (mm)          695482 non-nu

,Longitude (x),Latitude (y),Station Name,Climate ID,Date/Time,Year,Max Temp (°C),Min Temp (°C),Mean Temp (°C),Heat Deg Days (°C),Cool Deg Days (°C),Total Rain (mm),Total Snow (cm),Total Precip (mm),Snow on Grnd (cm),Dir of Max Gust (10s deg),Spd of Max Gust (km/h)
0,-64.51,44.83,AALDERSVILLE,8204020,2021-01-01,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,-64.51,44.83,AALDERSVILLE,8204020,2021-01-02,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,-64.51,44.83,AALDERSVILLE,8204020,2021-01-03,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,-64.51,44.83,AALDERSVILLE,8204020,2021-01-04,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,-64.51,44.83,AALDERSVILLE,8204020,2021-01-05,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [180]:
df.columns

Index(['Longitude (x)', 'Latitude (y)', 'Station Name', 'Climate ID',
       'Date/Time', 'Year', 'Max Temp (°C)', 'Min Temp (°C)', 'Mean Temp (°C)',
       'Heat Deg Days (°C)', 'Cool Deg Days (°C)', 'Total Rain (mm)',
       'Total Snow (cm)', 'Total Precip (mm)', 'Snow on Grnd (cm)',
       'Dir of Max Gust (10s deg)', 'Spd of Max Gust (km/h)'],
      dtype='str')

## ✅ Create Table with precipitation and snow
Create a new table with only these columns:

- Climate ID
- Date/Time
- Total Rain (mm)
- Total Snow (cm)
- Total Precip (mm)
- Snow on Grnd (cm)
- ✔ Keep the row if any one of these columns has a value
- ❌ Drop the row only if all four are NaN
Columns:

Total Rain (mm)
Total Snow (cm)
Total Precip (mm)
Snow on Grnd (cm)

✅ Only for dates from 1 Jan 2010 to 31 Dec 2025 (inclusive)

In [184]:
import pandas as pd

df["Date/Time"] = pd.to_datetime(df["Date/Time"])
df_2010_2025 = df[
    (df["Date/Time"] >= "2010-01-01") &
    (df["Date/Time"] <= "2025-12-31")
]
precip_2010_2025 = df_2010_2025[
    [
        "Climate ID",
        "Date/Time",
        "Total Rain (mm)",
        "Total Snow (cm)",
        "Total Precip (mm)",
        "Snow on Grnd (cm)"
    ]
]

In [186]:
precip_2010_2025.info()

<class 'pandas.DataFrame'>
Index: 662203 entries, 0 to 964632
Data columns (total 6 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   Climate ID         662203 non-null  str           
 1   Date/Time          662203 non-null  datetime64[us]
 2   Total Rain (mm)    110175 non-null  float64       
 3   Total Snow (cm)    108225 non-null  float64       
 4   Total Precip (mm)  473657 non-null  float64       
 5   Snow on Grnd (cm)  175541 non-null  float64       
dtypes: datetime64[us](1), float64(4), str(1)
memory usage: 35.4 MB


In [187]:
precip_2010_2025_clean = precip_2010_2025.dropna(
    subset=[
        "Total Rain (mm)",
        "Total Snow (cm)",
        "Total Precip (mm)",
        "Snow on Grnd (cm)"
    ],
    how="all"
)

In [188]:
precip_2010_2025_clean.info()

<class 'pandas.DataFrame'>
Index: 482572 entries, 253 to 964632
Data columns (total 6 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   Climate ID         482572 non-null  str           
 1   Date/Time          482572 non-null  datetime64[us]
 2   Total Rain (mm)    110175 non-null  float64       
 3   Total Snow (cm)    108225 non-null  float64       
 4   Total Precip (mm)  473657 non-null  float64       
 5   Snow on Grnd (cm)  175541 non-null  float64       
dtypes: datetime64[us](1), float64(4), str(1)
memory usage: 25.8 MB


In [189]:

precip_2010_2025_clean.to_csv(
    "precip_2010_2025.csv",
    index=False
)


## Create the three new DataFrames with the specified names

In [92]:
df_wind = df[
    [
        'Climate ID',
        'Date/Time',
        'Dir of Max Gust (10s deg)',
        'Spd of Max Gust (km/h)'
    ]
]

df_station = df[
    [
        'Longitude (x)',
        'Latitude (y)',
        'Station Name',
        'Climate ID'
    ]
]
df_temp = df[
    [
        'Climate ID',
        'Date/Time',
        'Max Temp (°C)',
        'Min Temp (°C)',
        'Mean Temp (°C)',
        'Heat Deg Days (°C)',
        'Cool Deg Days (°C)'
    ]
]

In [93]:
df_temp.info()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          964749 non-null  str    
 1   Date/Time           964749 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      826411 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 51.5 MB


##  Make df_station contain only unique Climate IDs (EDA for dim_Station)
- To make df_station contain only unique Climate IDs (one row per station), drop duplicate rows based on the Climate ID column.

In [94]:
df_station.info()
df_station = df_station.drop_duplicates(subset='Climate ID').reset_index(drop=True)


<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Longitude (x)  964749 non-null  float64
 1   Latitude (y)   964749 non-null  float64
 2   Station Name   964749 non-null  str    
 3   Climate ID     964749 non-null  str    
dtypes: float64(2), str(2)
memory usage: 29.4 MB


In [95]:
df_station.value_counts()

Longitude (x)  Latitude (y)  Station Name         Climate ID
-64.51         44.83         AALDERSVILLE         8204020       1
-54.00         47.31         ARGENTIA (AUT)       8400104       1
-67.72         46.71         AROOSTOOK            8100300       1
-65.47         43.45         BACCARO PT           8200255       1
-56.07         48.97         BADGER (AUT)         8400301       1
                                                               ..
-67.54         46.15         WOODSTOCK NEWBRIDGE  8105603       1
-59.31         47.71         WRECKHOUSE           8404343       1
-66.09         43.83         YARMOUTH A           8206495       1
                                                  8206496       1
                             YARMOUTH RCS         8206491       1
Name: count, Length: 136, dtype: int64

## EDA for dim_Station 
- compare your station list (df_station) with df_atlantic and then export the cleaned/compared station list to CSV.

In [96]:
df_stationlist = pd.read_csv("./../../../data/raw/Yashaswi/Stationlist_clean.csv")

In [97]:
df_stationlist.info()

<class 'pandas.DataFrame'>
RangeIndex: 137 entries, 0 to 136
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Station Name  137 non-null    str    
 1   Province      137 non-null    str    
 2   Latitude      137 non-null    float64
 3   Longitude     137 non-null    float64
 4   Climate ID    137 non-null    str    
 5   First Year    137 non-null    int64  
 6   Last Year     137 non-null    int64  
dtypes: float64(2), int64(2), str(3)
memory usage: 7.6 KB


**Identify Climate IDs NOT matching**
  - Climate IDs in df_station but missing from df_stationlist
  - Climate IDs in df_stationlist but missing from df_station
  - Quick counts (recommended sanity check)

In [98]:
not_in_clean = df_station[
    ~df_station['Climate ID'].isin(df_stationlist['Climate ID'])
]
not_in_station = df_stationlist[
    ~df_stationlist['Climate ID'].isin(df_station['Climate ID'])
]
print("In df_station but not in df_stationlist:",
      not_in_clean['Climate ID'].nunique())

print("In df_stationlist but not in df_station:",
      not_in_station['Climate ID'].nunique())


In df_station but not in df_stationlist: 0
In df_stationlist but not in df_station: 1


**1. Identify the mismatching Climate ID**

In [99]:
missing_ids = df_stationlist.loc[
    ~df_stationlist['Climate ID'].isin(df_station['Climate ID']),
    'Climate ID'
]

missing_ids


69    8502588
Name: Climate ID, dtype: str

**2. Check if this Climate ID exists in temperature data**

In [100]:
missing_climate_id = missing_ids.iloc[0]
df_temp_missing = df_temp[df_temp['Climate ID'] == missing_climate_id]


**3. Inspect the temperature values**

In [101]:
df_temp_missing.info()

<class 'pandas.DataFrame'>
RangeIndex: 0 entries
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Climate ID          0 non-null      str    
 1   Date/Time           0 non-null      str    
 2   Max Temp (°C)       0 non-null      float64
 3   Min Temp (°C)       0 non-null      float64
 4   Mean Temp (°C)      0 non-null      float64
 5   Heat Deg Days (°C)  0 non-null      float64
 6   Cool Deg Days (°C)  0 non-null      float64
dtypes: float64(5), str(2)
memory usage: 132.0 bytes


**Look for it in df_stationlist**

In [102]:
df_stationlist_info = df_stationlist[
    df_stationlist['Climate ID'] == missing_climate_id
]

df_stationlist_info



,Station Name,Province,Latitude,Longitude,Climate ID,First Year,Last Year
69,MAKKOVIK A,NEWFOUNDLAND AND LABRADOR,55.08,-59.19,8502588,2015,2026


In [103]:
df_station.info()

<class 'pandas.DataFrame'>
RangeIndex: 136 entries, 0 to 135
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Longitude (x)  136 non-null    float64
 1   Latitude (y)   136 non-null    float64
 2   Station Name   136 non-null    str    
 3   Climate ID     136 non-null    str    
dtypes: float64(2), str(2)
memory usage: 4.4 KB


# Finalize dim_station_clean

In [104]:
df_station.to_csv("dim_station_clean.csv", index=False)

✅ Final conclusion (important)
This Climate ID:

exists only in Stationlist_clean.csv
has no station metadata in my main dataset
has no temperature records

✅ Therefore, it is invalid for analysis
✅ It should be removed from the station list

## Temperature dim table EDA

In [105]:
df_temp.info()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          964749 non-null  str    
 1   Date/Time           964749 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      826411 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 51.5 MB


## Count NULL (NaN) values per column

In [106]:
temp_cols = [
    'Max Temp (°C)',
    'Min Temp (°C)',
    'Mean Temp (°C)',
    'Heat Deg Days (°C)',
    'Cool Deg Days (°C)'
]
null_counts = df_temp[temp_cols].isna().sum()
null_counts

Max Temp (°C)         136268
Min Temp (°C)         133632
Mean Temp (°C)        138338
Heat Deg Days (°C)    138338
Cool Deg Days (°C)    138338
dtype: int64

## 1. Find rows where ALL temperature columns are NULL

In [107]:
all_null_rows = df_temp[temp_cols].isna().all(axis=1)
all_null_count = all_null_rows.sum()
all_null_count

np.int64(131560)

## 2. Inspect those rows (important)

In [108]:
df_temp_all_null = df_temp[all_null_rows]
df_temp_all_null.head()
df_temp_all_null.info()

<class 'pandas.DataFrame'>
Index: 131560 entries, 0 to 964577
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          131560 non-null  str    
 1   Date/Time           131560 non-null  str    
 2   Max Temp (°C)       0 non-null       float64
 3   Min Temp (°C)       0 non-null       float64
 4   Mean Temp (°C)      0 non-null       float64
 5   Heat Deg Days (°C)  0 non-null       float64
 6   Cool Deg Days (°C)  0 non-null       float64
dtypes: float64(5), str(2)
memory usage: 8.0 MB


## Cleanup "all null" rows

In [109]:
df_temp_clean = df_temp[~all_null_rows].reset_index(drop=True)
df_temp_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 833189 entries, 0 to 833188
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          833189 non-null  str    
 1   Date/Time           833189 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      826411 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 44.5 MB


**Current situation (based on your output)

df_temp_all_null → ✅ rows where all temp columns were null (already isolated)
df_temp_clean → ✅ all‑null rows removed
Now inside df_temp_clean, there are:

✅ fully complete rows and Partially null rows**

## Find rows with any nulls (after all‑null rows are removed)
 -  Count partially‑null rows

In [110]:
partial_null_rows = df_temp_clean[temp_cols].isna().any(axis=1)
partial_null_count = partial_null_rows.sum()
partial_null_count

np.int64(6782)

In [111]:
df_temp_partial_null = df_temp_clean[partial_null_rows]

df_temp_partial_null.head()

,Climate ID,Date/Time,Max Temp (°C),Min Temp (°C),Mean Temp (°C),Heat Deg Days (°C),Cool Deg Days (°C)
1680,8400104,1995-01-23,NaN,-0.9,NaN,NaN,NaN
1687,8400104,1995-01-30,NaN,-6.2,NaN,NaN,NaN
1689,8400104,1995-02-02,NaN,-5.5,NaN,NaN,NaN
1695,8400104,1995-02-08,NaN,-7.9,NaN,NaN,NaN
1700,8400104,1995-02-13,-4.8,NaN,NaN,NaN,NaN


In [112]:
df_temp_partial_null.info()

<class 'pandas.DataFrame'>
Index: 6782 entries, 1680 to 830316
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Climate ID          6782 non-null   str    
 1   Date/Time           6782 non-null   str    
 2   Max Temp (°C)       2074 non-null   float64
 3   Min Temp (°C)       4710 non-null   float64
 4   Mean Temp (°C)      4 non-null      float64
 5   Heat Deg Days (°C)  4 non-null      float64
 6   Cool Deg Days (°C)  4 non-null      float64
dtypes: float64(5), str(2)
memory usage: 423.9 KB


# NEW RULE DEFINITION FOR ROW VALIDITY 
Remove rows that match this specific invalid pattern:
✅ Heat Deg Days and Cool Deg Days HAVE values
❌ Max Temp, Min Temp, and Mean Temp are ALL null
This is a very reasonable quality rule for climate data.

In [113]:
temp_cols_core = [
    'Max Temp (°C)',
    'Min Temp (°C)',
    'Mean Temp (°C)'
]

degday_cols = [
    'Heat Deg Days (°C)',
    'Cool Deg Days (°C)'
]
invalid_rows = (
    df_temp_clean[temp_cols_core].isna().all(axis=1) &      # all temps missing
    df_temp_clean[degday_cols].notna().any(axis=1)          # but deg days present
)
invalid_rows.sum()

np.int64(0)

# NEW FILLING RULE FOR ROW VALIDITY
✅ Filling rules (restated clearly)
For rows where Mean Temp (°C) is NA:


✅ If BOTH Max and Min are present
→ Mean = (Max + Min) / 2


✅ If ONLY ONE of Max or Min is present
→ Mean = the available value


❌ If BOTH Max and Min are missing
→ Leave Mean as NA (I have already removed fully-null rows)

In [114]:
df_t = df_temp_clean.copy()
mask_both = (
    df_t['Mean Temp (°C)'].isna() &
    df_t['Max Temp (°C)'].notna() &
    df_t['Min Temp (°C)'].notna()
)

df_t.loc[mask_both, 'Mean Temp (°C)'] = (
    df_t.loc[mask_both, ['Max Temp (°C)', 'Min Temp (°C)']].mean(axis=1)
)
mask_only_max = (
    df_t['Mean Temp (°C)'].isna() &
    df_t['Max Temp (°C)'].notna() &
    df_t['Min Temp (°C)'].isna()
)

df_t.loc[mask_only_max, 'Mean Temp (°C)'] = df_t.loc[
    mask_only_max, 'Max Temp (°C)'
]
mask_only_min = (
    df_t['Mean Temp (°C)'].isna() &
    df_t['Min Temp (°C)'].notna() &
    df_t['Max Temp (°C)'].isna()
)

df_t.loc[mask_only_min, 'Mean Temp (°C)'] = df_t.loc[
    mask_only_min, 'Min Temp (°C)'
]
df_t['Mean Temp (°C)'].isna().sum()

np.int64(0)

# Spot‑check correctness

In [115]:
df_t.loc[
    mask_both | mask_only_max | mask_only_min,
    ['Max Temp (°C)', 'Min Temp (°C)', 'Mean Temp (°C)']
].head(10)

,Max Temp (°C),Min Temp (°C),Mean Temp (°C)
1680,NaN,-0.9,-0.9
1687,NaN,-6.2,-6.2
1689,NaN,-5.5,-5.5
1695,NaN,-7.9,-7.9
1700,-4.8,NaN,-4.8
1710,NaN,-7.5,-7.5
1712,NaN,-0.2,-0.2
1723,NaN,0.9,0.9
1728,NaN,-6.6,-6.6
1730,NaN,-0.5,-0.5


# Finalize dim_temp data

In [116]:
df_temp_final = df_t
df_temp_final.info()

<class 'pandas.DataFrame'>
RangeIndex: 833189 entries, 0 to 833188
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          833189 non-null  str    
 1   Date/Time           833189 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      833189 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 44.5 MB


In [117]:
degday_cols = [
    'Heat Deg Days (°C)',
    'Cool Deg Days (°C)'
]
null_counts = df_temp_final[degday_cols].isna().sum()
null_counts

Heat Deg Days (°C)    6778
Cool Deg Days (°C)    6778
dtype: int64

In [118]:
zero_counts = (df_temp_final[degday_cols] == 0).sum()
zero_counts

Heat Deg Days (°C)     87776
Cool Deg Days (°C)    741738
dtype: int64

In [119]:
(
    (df_temp_final['Heat Deg Days (°C)'] > 0) &
    (df_temp_final['Cool Deg Days (°C)'] > 0)
).sum()


np.int64(0)

In [120]:
df_temp_final.insert(
    0,                      # position (first column)
    'temp_Index',                # column name
    range(1000, 1000 + len(df_temp_final))
)


In [121]:
df_temp_final.info()

<class 'pandas.DataFrame'>
RangeIndex: 833189 entries, 0 to 833188
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   temp_Index          833189 non-null  int64  
 1   Climate ID          833189 non-null  str    
 2   Date/Time           833189 non-null  str    
 3   Max Temp (°C)       828481 non-null  float64
 4   Min Temp (°C)       831117 non-null  float64
 5   Mean Temp (°C)      833189 non-null  float64
 6   Heat Deg Days (°C)  826411 non-null  float64
 7   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), int64(1), str(2)
memory usage: 50.9 MB


In [163]:
df_temp_final = df_temp_final.rename(columns={"Date/Time": "Date"})
df_temp_final = df_temp_final[
   (df_temp_final["Date"] <= "2025-12-31")
]
df_temp_final.head()

,temp_Index,Climate ID,Date,Max Temp (°C),Min Temp (°C),Mean Temp (°C),Heat Deg Days (°C),Cool Deg Days (°C)
0,1000,8204020,2021-09-11,18.7,8.7,13.7,4.3,0.0
1,1001,8204020,2021-09-12,22.4,9.8,16.1,1.9,0.0
2,1002,8204020,2021-09-13,20.8,11.0,15.9,2.1,0.0
3,1003,8204020,2021-09-14,17.2,6.9,12.1,5.9,0.0
4,1004,8204020,2021-09-15,20.2,5.7,13.0,5.0,0.0


In [165]:
df_temp_final.info()

<class 'pandas.DataFrame'>
Index: 818993 entries, 0 to 833072
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   temp_Index          818993 non-null  int64         
 1   Climate ID          818993 non-null  str           
 2   Date                818993 non-null  datetime64[us]
 3   Max Temp (°C)       814288 non-null  float64       
 4   Min Temp (°C)       816921 non-null  float64       
 5   Mean Temp (°C)      818993 non-null  float64       
 6   Heat Deg Days (°C)  812218 non-null  float64       
 7   Cool Deg Days (°C)  812218 non-null  float64       
dtypes: datetime64[us](1), float64(5), int64(1), str(1)
memory usage: 56.2 MB


In [166]:
df_temp_final.to_csv("temp_clean.csv", index=False)

In [167]:
df_temp_final["Date"].min(), df_temp_final["Date"].max()

(Timestamp('1995-01-01 00:00:00'), Timestamp('2025-12-31 00:00:00'))

In [168]:
df_temp_final["Date"] = pd.to_datetime(df_temp_final["Date"])
df_temp_post_2010 = df_temp_final[
   (( df_temp_final["Date"] >= "2010-01-01") & (df_temp_final["Date"] <= "2025-12-31"))
]

In [169]:
df_temp_post_2010["Date"].min(), df_temp_post_2010["Date"].max()

(Timestamp('2010-01-01 00:00:00'), Timestamp('2025-12-31 00:00:00'))

In [171]:
df_temp_post_2010.info()

<class 'pandas.DataFrame'>
Index: 572163 entries, 0 to 833072
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   temp_Index          572163 non-null  int64         
 1   Climate ID          572163 non-null  str           
 2   Date                572163 non-null  datetime64[us]
 3   Max Temp (°C)       569493 non-null  float64       
 4   Min Temp (°C)       571064 non-null  float64       
 5   Mean Temp (°C)      572163 non-null  float64       
 6   Heat Deg Days (°C)  568396 non-null  float64       
 7   Cool Deg Days (°C)  568396 non-null  float64       
dtypes: datetime64[us](1), float64(5), int64(1), str(1)
memory usage: 39.3 MB


In [170]:
df_temp_post_2010.to_csv("temp_clean_after2010.csv", index=False)

# EDA for dim_wind

In [125]:
df_wind.info()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 4 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Climate ID                 964749 non-null  str    
 1   Date/Time                  964749 non-null  str    
 2   Dir of Max Gust (10s deg)  463946 non-null  float64
 3   Spd of Max Gust (km/h)     469011 non-null  float64
dtypes: float64(2), str(2)
memory usage: 29.4 MB


## Null check

In [126]:

wind_cols = [
    'Dir of Max Gust (10s deg)',
    'Spd of Max Gust (km/h)'
]

df_wind[wind_cols].isna().sum()

Dir of Max Gust (10s deg)    500803
Spd of Max Gust (km/h)       495738
dtype: int64

In [127]:

any_null_rows = df_wind[wind_cols].isna().any(axis=1)
any_null_rows.sum()


np.int64(500803)

In [128]:

both_null_rows = df_wind[wind_cols].isna().all(axis=1)
both_null_rows.sum()


np.int64(495738)

# ✅ How to interpret this (important)


✅ Dir or Speed = NULL usually means:

No gust observed that day
Sensor limitation
Reporting rules (gusts often only logged when above threshold)

✅ Both NULL is very common in climatological wind datasets
→ not an error

⚠️ You should not drop these automatically unless your analysis requires gust data.

## For wind gust data: remove nulls logic

✅ REMOVE rows where both
Dir of Max Gust (10s deg) AND Spd of Max Gust (km/h) are NaN
✅ KEEP rows where at least one of them is present

This matches standard climatological practice.

In [129]:
df_wind_clean = df_wind[~both_null_rows].reset_index(drop=True)
df_wind_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 469011 entries, 0 to 469010
Data columns (total 4 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Climate ID                 469011 non-null  str    
 1   Date/Time                  469011 non-null  str    
 2   Dir of Max Gust (10s deg)  463946 non-null  float64
 3   Spd of Max Gust (km/h)     469011 non-null  float64
dtypes: float64(2), str(2)
memory usage: 14.3 MB


In [130]:
df_wind_clean = df_wind_clean.rename(columns={"Date/Time": "Date"})
df_wind_clean.head()

,Climate ID,Date,Dir of Max Gust (10s deg),Spd of Max Gust (km/h)
0,8204020,2021-09-11,28.0,43.0
1,8204020,2021-09-12,22.0,49.0
2,8204020,2021-09-13,31.0,34.0
3,8204020,2021-09-15,18.0,43.0
4,8204020,2021-09-16,23.0,33.0


In [131]:

df_wind_clean[
    ['Dir of Max Gust (10s deg)', 'Spd of Max Gust (km/h)']
].isna().all(axis=1).sum()


np.int64(0)

### dim_wind - Columns after cleaning

| Column                         | Non‑null count | Interpretation                                      |
|--------------------------------|----------------|-----------------------------------------------------|
| Climate ID                     | 469,011        | ✅ Intact                                            |
| Date/Time                      | 469,011        | ✅ Intact                                            |
| Dir of Max Gust (10s deg)      | 463,946        | Some gusts lack direction (normal)                  |
| Spd of Max Gust (km/h)         | 469,011        | Every remaining row has gust speed                  |

In [132]:
df_wind_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 469011 entries, 0 to 469010
Data columns (total 4 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Climate ID                 469011 non-null  str    
 1   Date                       469011 non-null  str    
 2   Dir of Max Gust (10s deg)  463946 non-null  float64
 3   Spd of Max Gust (km/h)     469011 non-null  float64
dtypes: float64(2), str(2)
memory usage: 14.3 MB


In [133]:
# Convert Date column to datetime
df_wind_clean["Date"] = pd.to_datetime(df_wind_clean["Date"])

# Filter date range
df_wind_clean = df_wind_clean[
    (df_wind_clean["Date"] >= "2015-01-01") &
    (df_wind_clean["Date"] <= "2026-12-31")
]

In [134]:
df_wind_clean.info()

<class 'pandas.DataFrame'>
Index: 302798 entries, 0 to 469010
Data columns (total 4 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   Climate ID                 302798 non-null  str           
 1   Date                       302798 non-null  datetime64[us]
 2   Dir of Max Gust (10s deg)  299563 non-null  float64       
 3   Spd of Max Gust (km/h)     302798 non-null  float64       
dtypes: datetime64[us](1), float64(2), str(1)
memory usage: 11.6 MB


In [135]:

df_wind_clean.insert(
    0,                      # position (first column)
    'Index',                # column name
    range(100, 100 + len(df_wind_clean))
)


In [136]:
df_wind_clean.info()

<class 'pandas.DataFrame'>
Index: 302798 entries, 0 to 469010
Data columns (total 5 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   Index                      302798 non-null  int64         
 1   Climate ID                 302798 non-null  str           
 2   Date                       302798 non-null  datetime64[us]
 3   Dir of Max Gust (10s deg)  299563 non-null  float64       
 4   Spd of Max Gust (km/h)     302798 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(1)
memory usage: 13.9 MB


In [137]:
(df_wind_clean["Date"] > "2025-12-31").sum()

np.int64(10424)

In [138]:
df_wind_clean = df_wind_clean[df_wind_clean["Date"] <= "2025-12-31"]

In [139]:
df_wind_clean.head()

,Index,Climate ID,Date,Dir of Max Gust (10s deg),Spd of Max Gust (km/h)
0,100,8204020,2021-09-11,28.0,43.0
1,101,8204020,2021-09-12,22.0,49.0
2,102,8204020,2021-09-13,31.0,34.0
3,103,8204020,2021-09-15,18.0,43.0
4,104,8204020,2021-09-16,23.0,33.0


In [140]:
df_wind_clean[
    ["Dir of Max Gust (10s deg)", "Spd of Max Gust (km/h)"]
].isna().any()

Dir of Max Gust (10s deg)     True
Spd of Max Gust (km/h)       False
dtype: bool

In [141]:
df_wind_clean.info()

<class 'pandas.DataFrame'>
Index: 292374 entries, 0 to 468914
Data columns (total 5 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   Index                      292374 non-null  int64         
 1   Climate ID                 292374 non-null  str           
 2   Date                       292374 non-null  datetime64[us]
 3   Dir of Max Gust (10s deg)  289141 non-null  float64       
 4   Spd of Max Gust (km/h)     292374 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(1)
memory usage: 13.4 MB


In [142]:
df_wind_clean = df_wind_clean.sort_values(by="Date", ascending=True)

In [144]:
df_wind_clean["wind_key"] = range(100, 100 + len(df_wind_clean))
cols = ["wind_key"] + [c for c in df_wind_clean.columns if c != "wind_key"]
df_wind_clean = df_wind_clean[cols]

In [145]:
df_wind_clean.head()

,wind_key,Index,Climate ID,Date,Dir of Max Gust (10s deg),Spd of Max Gust (km/h)
357788,100,227495,8403619,2015-01-01,21.0,33.0
179381,101,106839,8202000,2015-01-01,23.0,32.0
413116,102,265472,8205702,2015-01-01,22.0,33.0
455589,103,290845,8404343,2015-01-01,18.0,56.0
249640,104,157286,8102848,2015-01-01,27.0,33.0


In [148]:
df_wind_clean = df_wind_clean.drop(columns=["Index"])

KeyError: "['Index'] not found in axis"

In [149]:
df_wind_clean = df_wind_clean.reset_index(drop=True)

In [152]:
df_wind_clean.head()

,wind_key,Climate ID,Date,Dir of Max Gust (10s deg),Spd of Max Gust (km/h)
0,100,8403619,2015-01-01,21.0,33.0
1,101,8202000,2015-01-01,23.0,32.0
2,102,8205702,2015-01-01,22.0,33.0
3,103,8404343,2015-01-01,18.0,56.0
4,104,8102848,2015-01-01,27.0,33.0


In [151]:
df_wind_clean.index[:5]

RangeIndex(start=0, stop=5, step=1)

In [153]:
df_wind_clean.to_csv("dim_wind_clean.csv", index=False)